src/server/pfs/db/driver.go

```diff
@@ -0,0 +1,1021 @@
+		return err
+	}
+
+	return nil
+}
+
+func dbConnect(address string) (*gorethink.Session, error) {
+	return gorethink.Connect(gorethink.ConnectOpts{
+		Address: address,
+		Timeout: connectTimeoutSeconds * time.Second,
+	})
+}
+
+func validateRepoName(name string) error {
+	match, _ := regexp.MatchString("^[a-zA-Z0-9_]+$", name)
+
+	if !match {
+		return fmt.Errorf("repo name (%v) invalid: only alphanumeric and underscore characters allowed", name)
+	}
+
+	return nil
+}
+
+func (d *driver) getTerm(table Table) gorethink.Term {
+	return gorethink.DB(d.dbName).Table(table)
+}
+
+func (d *driver) CreateRepo(repo *pfs.Repo, created *google_protobuf.Timestamp,
+	provenance []*pfs.Repo, shards map[uint64]bool) error {
+
+	err := validateRepoName(repo.Name)
+	if err != nil {
+		return err
+	}
+
+	_, err = d.getTerm(repoTable).Insert(&persist.Repo{
+		Name:    repo.Name,
+		Created: created,
+	}).RunWrite(d.dbClient)
+	return err
+}
+
+func (d *driver) InspectRepo(repo *pfs.Repo, shards map[uint64]bool) (repoInfo *pfs.RepoInfo, retErr error) {
+	cursor, err := d.getTerm(repoTable).Get(repo.Name).Run(d.dbClient)
+	if err != nil {
+		return nil, err
+	}
+	rawRepo := &persist.Repo{}
+	if err := cursor.One(rawRepo); err != nil {
+		return nil, err
+	}
+	repoInfo = &pfs.RepoInfo{
+		Repo:    &pfs.Repo{rawRepo.Name},
+		Created: rawRepo.Created,
+	}
+	return repoInfo, nil
+}
+
+func (d *driver) ListRepo(provenance []*pfs.Repo, shards map[uint64]bool) (repoInfos []*pfs.RepoInfo, retErr error) {
+	cursor, err := d.getTerm(repoTable).Run(d.dbClient)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := cursor.Close(); err != nil && retErr == nil {
+			retErr = err
+		}
+	}()
+	for {
+		repo := &persist.Repo{}
+		if !cursor.Next(repo) {
+			break
+		}
+		repoInfos = append(repoInfos, &pfs.RepoInfo{
+			Repo:    &pfs.Repo{repo.Name},
+			Created: repo.Created,
+		})
+	}
+	if err := cursor.Err(); err != nil {
+		return nil, err
+	}
+	return repoInfos, nil
+}
+
+func (d *driver) DeleteRepo(repo *pfs.Repo, shards map[uint64]bool, force bool) error {
+	_, err := d.getTerm(repoTable).Get(repo.Name).Delete().RunWrite(d.dbClient)
+	return err
+}
+
+func (d *driver) StartCommit(repo *pfs.Repo, commitID string, parentID string, branch string, started *google_protobuf.Timestamp, provenance []*pfs.Commit, shards map[uint64]bool) (retErr error) {
+	var _provenance []string
+	for _, c := range provenance {
+		_provenance = append(_provenance, c.ID)
+	}
+	commit := &persist.Commit{
+		ID:         commitID,
+		Repo:       repo.Name,
+		Started:    prototime.TimeToTimestamp(time.Now()),
+		Provenance: _provenance,
+	}
+
+	if branch == "" {
+		branch = uuid.NewWithoutDashes()
+	}
+
+	var clockID *persist.ClockID
+	if parentID == "" {
+		for {
+			// The head of this branch will be our parent commit
+			parentCommit := &persist.Commit{}
+			err := d.getHeadOfBranch(repo.Name, branch, parentCommit)
```

Comment:
This could be a potential race if 2 commits come in on the same branch at the same time couldn't it? In that case it seems like we'd wind up with 2 commits with the same branch clock wouldn't we?